Strengthen thread safety assertions

ajtowns opened the pull request:
This changes `LOCK(mutex)` for non-global, non-recursive mutexes to be annotated with the negative capability for the mutex it refers to, to prevent . clang applies negative capabilities recursively, so this helps avoid forgetting to annotate functions.

This can't reasonably be used for globals, because clang would require every function to be annotated with `EXCLUSIVE_LOCKS_REQUIRED(!g_mutex)` for each global mutex; so this introduces a trivial `GlobalMutex` subclass of `Mutex`, and reduces the annotations for both `GlobalMutex`  to `LOCKS_EXCLUDED` which only catches trivial errors (eg (`LOCK(x); LOCK(x);`).

ajtowns committed:
sync.h: Add GlobalMutex type

ajtowns committed:
scripted-diff: Convert global Mutexes to GlobalMutexes

-BEGIN VERIFY SCRIPT-
sed -i -E -e '/^([a-z]+ )?Mutex [a-z]/ s/Mutex/GlobalMutex/' $(git grep -lE '^([a-z]+ )?Mutex [a-z]')
-END VERIFY SCRIPT-

ajtowns committed:
sync.h: Imply negative assertions when calling LOCK

ajtowns committed:
doc: Update developer notes

ajtowns commented:
This hopefully avoids bugs like https://github.com/bitcoin/bitcoin/pull/24157#discussion_r849698742 and gets most of the benefits of `-Wthread-safety-negative` without introducing massive amounts of unhelpful warnings about `cs_main` and the like as per #20272. See also #21598.

jonatack commented:
Concept ACK

jamesob commented:
Concept ACK

hebasto commented:
Concept ACK. I had a similar perspective view since f8213c05f087e5fbb5d92a291f766b0baebc798f :)

Maybe split of changes related to `GlobalMutex` as they looks non-controversial and pretty straightforward to review?

jonatack commented on src/checkqueue.h:
```diff
@@ -188,7 +188,7 @@ class CCheckQueue
 
     //! Stop all of the worker threads.
-    void StopWorkerThreads()
+    void StopWorkerThreads() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
```
Should each of the annotations added in the first commit also have `AssertLockNotHeld`?

ajtowns replied:
That feels like overkill, but I'm not sure where to draw the line.

MarcoFalke replied:
Closing as resolved?

jonatack commented on src/sync.h:
```diff
@@ -129,7 +127,25 @@ class LOCKABLE AnnotatedMixin : public PARENT
+ *
+ * Thread safety analysis can't handle negative assertions about mutex's
+ * with global scope well, so mark them with a separate type, and
+ * eventually move all the mutexes into classes so they are not globally
```
nit, "mutexes" here is correct, but s/Mutex's/Mutexes/ on line 130 and s/mutex's/mutexes/ on line 134

jonatack reviewed (commented):
Is the goal of this change to allow building with `-Wthread-safety-negative`? Still seeing many warnings on this branch with `./configure CC=clang CXX=clang++ CXXFLAGS='-Wthread-safety -Wthread-safety-negative' && make clean && make`. A regular debug build on this branch is clean, however.

ajtowns commented:
> Is the goal of this change to allow building with `-Wthread-safety-negative`?

No; as far as I can see that option prevents you from holding a `RecursiveMutex` when reacquiring the same `RecursiveMutex`, so is pointless until we don't have any of those left.

w0xlt commented:
Concept ACK. Will review soon.

ajtowns commented:
Rebased past #24157, fixed bad apostrophes

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/24931/commits/e735d93ce456eb37898ae11e623702e46970b80f

I changed a `RecursiveMutex` to `Mutex` and the compilation failed with warnings `requires negative capability` for `AssertLockNotHeld`, `LOCK` and `LOCK2`.

vasild commented on src/versionbits.h:
```diff
@@ -87,21 +87,21 @@ class VersionBitsCache
      * If provided, signalling_blocks is set to true/false based on whether each block in the period signalled
      */
-    static BIP9Stats Statistics(const CBlockIndex* pindex, const Consensus::Params& params, Consensus::DeploymentPos pos, std::vector<bool>* signalling_blocks = nullptr);
+    static BIP9Stats Statistics(const CBlockIndex* pindex, const Consensus::Params& params, Consensus::DeploymentPos pos, std::vector<bool>* signalling_blocks = nullptr) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
```
This change seems unnecessary because this method does not lock any mutex or call a method that requires `!m_mutex`:

```cpp
BIP9Stats VersionBitsCache::Statistics(const CBlockIndex* pindex, const Consensus::Params& params, Consensus::DeploymentPos pos, std::vector<bool>* signalling_blocks)
{
    return VersionBitsConditionChecker(pos).GetStateStatisticsFor(pindex, params, signalling_blocks);
}   
```

ajtowns replied:
It's also static, so seems like it would have a hard time finding an `m_mutex` to reference...

vasild commented on src/sync.h:
```diff
@@ -129,7 +127,25 @@ class LOCKABLE AnnotatedMixin : public PARENT
+
+#define AssertLockHeld(cs) AssertLockHeldInternal(#cs, __FILE__, __LINE__, &cs)
+
+inline void AssertLockNotHeldInline(const char* pszName, const char* pszFile, int nLine, Mutex* cs) EXCLUSIVE_LOCKS_REQUIRED(!cs) { AssertLockNotHeldInternal(pszName, pszFile, nLine, cs); }
```
style: `AssertLockNotHeldInline()` is newly added function, so should abide to the naming convention for its arguments:

```suggestion
inline void AssertLockNotHeldInline(const char* mutex_name, const char* file_name, int line, Mutex* mutex) EXCLUSIVE_LOCKS_REQUIRED(!cs) { AssertLockNotHeldInternal(mutex_name, file_name, line, mutex); }
```

vasild reviewed (commented):
ACK 314d2d4879 - just the first two commits, modulo the inline comments below.

The 3rd commit adds `GlobalMutex` type and having two types: `Mutex` and `GlobalMutex` seems like [one more "WTF" moment](http://www.osnews.com/images/comics/wtfm.jpg) to me.

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/24931/commits/c7a011c0d18206ca79e54b9cfef3f6abce456889

ajtowns commented:
Rebased past #24543

ajtowns commented:
Rebased past #25100 and split the first two commits out into #25109; please review that one first, it should be easy.

hebasto commented:
It seems the `g_shutdown_mutex` has been missed to be converted from `Mutex` into `GlobalMutex`, no?

hebasto commented on src/sync.h:
```diff
@@ -232,12 +244,26 @@ class SCOPED_LOCKABLE UniqueLock : public Base
+inline Mutex& MaybeCheckNotHeld(Mutex& cs) EXCLUSIVE_LOCKS_REQUIRED(!cs) LOCK_RETURNED(cs) { return cs; }
+inline Mutex* MaybeCheckNotHeld(Mutex* cs) EXCLUSIVE_LOCKS_REQUIRED(!cs) LOCK_RETURNED(cs) { return cs; }
+
+// When locking a GlobalMutex, just check we don't know it's locked
```
Sorry for bothering, but it's hard for me to parse this comment (sure I need to improve my English).

Maybe, re-word it a bit?

MarcoFalke replied:
```suggestion
// When locking a GlobalMutex, just check it is not locked in the surrounding scope.
```

ajtowns replied:
Done; thanks @MarcoFalke for the suggestion so I didn't have to ask for one :)

hebasto reviewed (commented):
Approach ACK e6f05cc7dfa021a0aef03d1ae904aa3b9555b1aa.

ajtowns commented on doc/developer-notes.md:
```diff
@@ -902,14 +902,19 @@ Threads and synchronization
 - Consistently use [Clang Thread Safety Analysis](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html) annotations to
-  get compile-time warnings about potential race conditions in code. Combine annotations in function declarations with
-  run-time asserts in function definitions:
+  get compile-time warnings about potential race conditions in code.
```
"get compile-time warnings about potential race conditions and deadlocks in code" ?

ajtowns replied:
Done

hebasto committed:
refactor: Propagate negative `!m_most_recent_block_mutex` capability

Could be verified with
$ ./configure CC=clang CXX=clang++ CXXFLAGS='-Wthread-safety -Wthread-safety-negative'
$ make clean
$ make 2>&1 | grep m_most_recent_block_mutex

hebasto committed:
refactor: Propagate negative `!m_tx_relay_mutex` capability

Could be verified with
$ ./configure CC=clang CXX=clang++ CXXFLAGS='-Wthread-safety -Wthread-safety-negative'
$ make clean
$ make 2>&1 | grep m_tx_relay_mutex

hebasto commented:
Looks like this PR has a silent merge conflict because of #22778 and #24062.

See #25175.

hebasto commented:
> This changes `LOCK(mutex)` for non-global, non-recursive mutexes to be annotated with the negative capability for the mutex it refers to, to prevent . clang applies negative capabilities recursively, so this helps avoid forgetting to annotate functions.

How about `WITH_LOCK` macro?

ajtowns commented:
Rebased onto #25175. The `WITH_LOCK` macro is based on `LOCK`, so also enforces these checks. `LOCK2`, `TRY_LOCK` and `WAIT_LOCK` are also updated.

hebasto commented:
> Rebased onto #25175. The `WITH_LOCK` macro is based on `LOCK`, so also enforces these checks. `LOCK2`, `TRY_LOCK` and `WAIT_LOCK` are also updated.

https://github.com/bitcoin/bitcoin/blob/cd6ae6906602090a1f3537e9abf97e25e4a6421a/src/net_processing.cpp#L294-L297

does not emit warning about missed `EXCLUSIVE_LOCKS_REQUIRED(!m_tx_relay_mutex)` annotation.

ajtowns commented:
Though, on the other hand, `WITH_LOCK` is within a lambda (in order to be able to `return` the result of the expression) and the annotations aren't propogated to the lambda. I think this is probably okay: `RecursiveMutex` and `GlobalMutex` don't require propogation anyway, and `Mutex` are all within a class so the new lambda will be implicitly tagged with the negative capability anyway?

However, tagging the `WITH_LOCK` lambda with `LOCKS_EXCLUDED(cs)` could be done, just as a double check that it's not being used when the lock is already held?

ajtowns committed:
net_processing: thread safety annotation for m_tx_relay_mutex

ajtowns committed:
qt/clientmodel: thread safety annotation for m_cached_tip_mutex

ajtowns commented:
> https://github.com/bitcoin/bitcoin/blob/cd6ae6906602090a1f3537e9abf97e25e4a6421a/src/net_processing.cpp#L294-L297
> 
> does not emit a warning about missed `EXCLUSIVE_LOCKS_REQUIRED(!m_tx_relay_mutex)` annotation.

Updated with negative assertions propogated when WITH_LOCK is used, which caught the above case and another. Couldn't see a way to avoid having WITH_LOCK expand cs twice though.

hebasto commented on src/sync.h:
```diff
@@ -129,10 +129,22 @@ using RecursiveMutex = AnnotatedMixin<std::recursive_mutex>;
+ *
+ * See: https://github.com/bitcoin/bitcoin/pull/20272#issuecomment-720755781
+ */
+class GlobalMutex : public Mutex { };
```
a559509a0b8cade27199740212d7b589f71a0e3b

[`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy) complains:
```suggestion
class GlobalMutex : public Mutex
{
};
```

hebasto reviewed (approved):
ACK ce893c0497fc9b8ab9752153dfcc77c9f427545e

vasild commented on src/sync.h:
```diff
@@ -276,7 +302,7 @@ using DebugLock = UniqueLock<typename std::remove_reference<typename std::remove
 //!
 //! The above is detectable at compile-time with the -Wreturn-local-addr flag in
 //! gcc and the -Wreturn-stack-address flag in clang, both enabled by default.
-#define WITH_LOCK(cs, code) [&]() -> decltype(auto) { LOCK(cs); code; }()
+#define WITH_LOCK(cs, code) (MaybeCheckNotHeld(cs), [&]() -> decltype(auto) { LOCK(cs); code; }())
```
While playing with this I observed that it is possible to circumvent the thread safety warnings mechanism by wrapping the lock call in a lambda and calling this lambda immediately. That circumvention only works for mutexes that are a member of a class:

```cpp
Mutex mutex_global;

// Commented annotations would silence the warnings, clang 15
class C
{
public:
    Mutex mutex_member;

    void f1() // EXCLUSIVE_LOCKS_REQUIRED(!mutex_member)
    {
        // warning: calling function 'MaybeCheckNotHeld' requires negative capability '!mutex_member' [-Wthread-safety-analysis]
        LOCK(mutex_member);
    }

    void f2()
    {
        [&]() { LOCK(mutex_member); }(); // no need for annotations
    }

    void f3() // EXCLUSIVE_LOCKS_REQUIRED(!mutex_global)
    {
        // warning: calling function 'MaybeCheckNotHeld' requires negative capability '!mutex_global' [-Wthread-safety-analysis]
        LOCK(mutex_global);
    }

    void f4() // EXCLUSIVE_LOCKS_REQUIRED(!mutex_global)
    {
        // warning: calling function 'MaybeCheckNotHeld' requires negative capability '!mutex_global' [-Wthread-safety-analysis]
        [&]() /* EXCLUSIVE_LOCKS_REQUIRED(!mutex_global) */ { LOCK(mutex_global); }();
    }
};
```

Looks like the construct used for `WITH_LOCK()` in this PR relies on the `f2()` behavior from the example above. Is that behavior documented somewhere in the clang docs? If not, then it may be wiser to not depend on undocumented behavior.

ajtowns replied:
You can avoid the thread safety warnings in a lot of ways if you create aliases for either the locks or the variables being guarded; eg:

```c++
    int i GUARDED_BY(mutex_member) = 0;
    void f5()
    {
        int* i_alias = &i;
        *i_alias = 5;
    }
```

(Using `[&]` here is creating an implicit alias for all the member variables, though clang does a better job of analysing past the aliases there than in the above example...)

I think the best long term approach is to do away with any uses of both `GlobalMutex` and `RecursiveMutex`, at which point we can annotate the lambda in `WITH_LOCK` with `EXCLUSIVE_LOCKS_REQUIRED(!cs)` and we're done. Having it work in the meantime allows us to be more confident that we don't have any potential recursive cases when switching a `RecursiveMutex` to a regular `Mutex`, so seems like it aids that goal to me.

vasild replied:
Ok, this explains why `f2()` needs no annotations. I guess at some point the compiler may become smarter, detect the reference and issue a warning.

> I think the best long term approach is to do away with any uses of both `GlobalMutex` and `RecursiveMutex`, at which point we can annotate the lambda in `WITH_LOCK` with `EXCLUSIVE_LOCKS_REQUIRED(!cs)` and we're done.

I agree. Do you envision that we do not have mutexes in global scope, or we have ones that are of type `Mutex`?

ajtowns replied:
Anytime you currently have:

```c++
GlobalMutex g_mut;
int g_importantnum GUARDED_BY(g_mut) = 0;
std::string g_importantstr GUARDED_BY(g_mut);

void bump_importantint(int offset)
{
    LOCK(g_mut);
    g_importantnum += offset;
}
```

you can rewrite it as

```c++
class important_stuff
{
private:
    Mutex m_mut;
    int m_num GUARDED_BY(m_mut) = 0;
    std::string m_str GUARDED_BY(m_mut);
public:
    void bump(int offset) EXCLUSIVE_LOCKS_REQUIRED(!m_mut);
};
important_stuff g_important;
```

or similar (depending on whether you can move all the code that uses the mutex into the class, or if you have to expose the mutex), which doesn't require very large changes to the code, but does make for a better fit with the thread safety analysis. Might also make it easy to later change `g_important` to not need to be a global too.

vasild replied:
> Do you envision that we do not have mutexes in global scope, or we have ones that are of type `Mutex`

I guess your reply above favors the "we do not have mutexes in global scope" case, right?

MarcoFalke commented:
Maybe drop the merged commits from the broken GitHub web view?


```
git commit --allow-empty -m empty
git push ...
git reset --hard HEAD~
git push ... --force

MarcoFalke commented:
Nvm, the merge commit will have the correct commit list.

review ACK ce893c0497fc9b8ab9752153dfcc77c9f427545e 🐦

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK ce893c0497fc9b8ab9752153dfcc77c9f427545e 🐦
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjrDgwAuuYRw/pMLEzZ0bmQYj83qnnVrTospYbV67FoYgA4pfE9zHVJ6eN25SPP
akqhk9h1mhKVESBkBYr/7mrgvJjCqA7+TUBR5RWuQKrvu54JKxMTgEMyJ9nETT4a
iQKqbVU8TAcAR84A9nYFgQdwGF4OP1xm4SthwIiAEFw9aNeUPWRMuPQc7OoB+3/i
J6yjD6pI+TBLIYCSqK232QLw5kSR+m17GY4lg5akpdZi61o8jJ4ieGJKDZf8G54O
XHT2UR0+dJVpkWQFDuseYbXUsVVe8lhXPNwKCbhsvzhVsJueAeq0LVhHfDtydlQC
OKA4qQ2hxPr5O2PMJANv52QXyDs8InWj0+0bqOr/MDvtiZd+hdLhv9lS+PLznnLx
r4pLcQv8esvPya/erbcM0TCp442MjwMadD1V91eEl3L7VWDQ5gFxozimGncaIL2u
h5WW7VchbrYfJVna7KtOP6Hpey3aSfKP6Extba4jnz0mEYnXHut1Kc3S8K0+TwB+
PG/M6vik
=Pc8D
-----END PGP SIGNATURE-----
```


</details>
